### ISSUE TITLE
[SmtpClient] SendAsync/SendMailAsync fails with DeliveryFormat=SmtpDeliveryFormat.SevenBit

### ISSUE THREAD

**Issue by xiangyuecn** [dotnet/corefx#33480]
## Problem

When sending mail using the asynchronous method (`SendAsync`, `SendMailAsync`), set `DeliveryFormat` `=` `SmtpDeliveryFormat.SevenBit`，If mail server supports `SMTPUTF8`, the configuration will be invalid. The UTF-8 contents(All Character) in `Subject`、`Attachments file name`, will be sent directly in the form of `UTF-8 encoding` plain text, `Base64` transcoding will not be performed. The synchronization method (`Send`) is not affected by this.

Specific details refer to questions：[https://developercommunity.visualstudio.com/content/problem/381046/smtpclient%E4%B8%80%E5%A4%84%E4%BB%A3%E7%A0%81%E7%BC%96%E5%86%99%E9%94%99%E8%AF%AF%E5%AF%BC%E8%87%B4%E5%BC%82%E6%AD%A5%E5%8F%91%E9%80%81%E9%82%AE%E4%BB%B6%E6%97%B6deliveryformat%E9%85%8D%E7%BD%AE%E9%A1%B9%E6%97%A0%E6%B3%95%E6%AD%A3%E7%A1%AE%E5%B7%A5%E4%BD%9C.html](https://developercommunity.visualstudio.com/content/problem/381046/smtpclient%E4%B8%80%E5%A4%84%E4%BB%A3%E7%A0%81%E7%BC%96%E5%86%99%E9%94%99%E8%AF%AF%E5%AF%BC%E8%87%B4%E5%BC%82%E6%AD%A5%E5%8F%91%E9%80%81%E9%82%AE%E4%BB%B6%E6%97%B6deliveryformat%E9%85%8D%E7%BD%AE%E9%A1%B9%E6%97%A0%E6%B3%95%E6%AD%A3%E7%A1%AE%E5%B7%A5%E4%BD%9C.html)

## Reappear code

Note：I use .NET Framework 4.5 to test the code, Because the affected part of the .NET Framework is consistent with .NET Core and does not affect replication, but I did not test .NET Core.
``` C#
var yourOutlookMailAddr = "xiangyuecn@outlook.com";//This is changed to your outlook mailbox. Outlook supports SMTPUTF8.

var msg = new MailMessage("test@localhost", yourOutlookMailAddr);
msg.HeadersEncoding = msg.BodyEncoding = msg.SubjectEncoding = Encoding.UTF8;

var smtp = new SmtpClient("outlook-com.olc.protection.outlook.com", 25);
smtp.DeliveryFormat = SmtpDeliveryFormat.SevenBit;

try {
	//Affected transmission mode
	msg.Subject = "SendMailAsync 中文 Test 测试";
	var task = smtp.SendMailAsync(msg);
	task.Wait();
	Console.WriteLine("SendMailAsync OK");

	//Unaffected mode of transmission
	msg.Subject = "Send 中文 Test 测试";
	smtp.Send(msg);
	Console.WriteLine("Send OK");
} catch (AggregateException e) {
	Console.WriteLine("Send Fail：");
	Console.WriteLine(e.InnerException.ToString());
}
Console.ReadLine();
```

### inspect
You can use Wireshark to find that the `Subject Header` content sent by `SendMailAsync` is not encoded at all. The Base64 code is sent by the `Send` method.

In addition, you can go to the outlook mailbox (** garbage box **) to check the source of the mail, and you will find it:

`SendMailAsync` Method sent (this is wrong):
```
Subject: SendMailAsync ä¸­æ Test æµè¯
```

`Send` Method sent (this is right):
```
Subject: =?utf-8?B?U2VuZCDkuK3mlocgVGVzdCDmtYvor5U=?=
```

## Causes and Solutions

The root cause is that in the `SmtpClient.SendMailCallback` method the `message.BeginSend` `allowUnicode` parameter is passed directly into `ServerSupportsEai`, instead of a unified `IsUnicodeSupported()`.

https://github.com/dotnet/corefx/blob/d68d72fe6f53d30e47168b7fd5aa664c6b3d9f6d/src/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs#L965-L968

Change `ServerSupportsEai` to `IsUnicodeSupported()` problem solving.
```c#
// src/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs

private void SendMailCallback(IAsyncResult result)
	...
	_message.BeginSend(_writer, DeliveryMethod != SmtpDeliveryMethod.Network,
                        IsUnicodeSupported(), new AsyncCallback(SendMessageCallback), result.AsyncState);
	...
}
```

## Found the affected version

- .NET Framework 4.5 - 4.7.2
- .NET Core Latest version

**Comment by davidsh** [comment]
>I think the behavior is allowed per RFC.

I think the net result of our behavior in SmtpClient is likely incorrect if the Subject is being displayed incorrectly in most email client (such as Outlook) etc.  

>Subject: SendMailAsync ä¸­æ�� Test æµ�è¯�

If this is what is being displayed in Outlook, for example, then SmtpClient is behaving incorrectly and should be fixed.

**Comment by davidsh** [comment]
Looks like PR #33793 is trying to address this problem.  That PR should be reviewed and additional testing should be done on it (perhaps manually if necessary) to verify correct behavior.

**Comment by xiangyuecn** [comment]
@caesar1995 Reject this interpretation. Because if you're right about that, there must be one that doesn't meet the criteria.

Results of `Send` and `SendAsync`, which is correct? :see_no_evil:

**Comment by hyzx86** [comment]
Has this patch been released yet?
Will it take effect in the latest version or only a few specified versions

**Comment by karelz** [comment]
It was fixed in 3.0, so you should see it in latest previews.
